File: backend/app/api/purchases/company_party_line.py
"""Форматирование строк поставщика/покупателя для счёта (§5 ТЗ)."""


def strip_leading_org_type(company_name: str, company_type: str | None) -> str:
	name = (company_name or "").strip()
	org_type = (company_type or "").strip()
	if not org_type or not name:
		return name
	if name.upper().startswith(org_type.upper()):
		return name[len(org_type) :].strip()
	return name


def format_legal_address_for_party(index: str | None, legal_address: str | None) -> str:
	addr = (legal_address or "").strip()
	idx = (index or "").strip()
	if not addr:
		return idx
	if idx and (addr.startswith(idx) or addr.startswith(f"{idx},")):
		return addr
	if idx:
		return f"{idx}, {addr}"
	return addr


def format_company_party_line(party: dict) -> str:
	name = party.get("company_name") or ""
	if not name:
		return ""
	org_type = (party.get("company_type") or "").strip()
	short_name = strip_leading_org_type(name, org_type or None)
	full_name = f"{org_type} {short_name}".strip() if org_type else short_name
	address = format_legal_address_for_party(party.get("index"), party.get("legal_address"))
	parts = [full_name, party.get("inn", ""), party.get("kpp", ""), address]
	return ", ".join(str(p).strip() for p in parts if p not in (None, ""))

File: frontend/utils/companyPartyLine.ts
/** Убирает тип организации из начала названия, если он уже там есть (ООО + «ООО …»). */
export function stripLeadingOrgType(companyName: string, companyType?: string): string {
	const name = companyName.trim()
	const type = companyType?.trim()
	if (!type || !name) return name
	if (name.toUpperCase().startsWith(type.toUpperCase())) {
		return name.slice(type.length).trim()
	}
	return name
}

/** Адрес для строки реквизитов: индекс не дублируем, если уже в legalAddress. */
export function formatLegalAddressForParty(index?: string, legalAddress?: string): string {
	const addr = (legalAddress ?? '').trim()
	const idx = (index ?? '').trim()
	if (!addr) return idx
	if (idx && (addr.startsWith(idx) || addr.startsWith(`${idx},`))) {
		return addr
	}
	if (idx) return `${idx}, ${addr}`
	return addr
}

export type PartyLineFields = {
	companyType?: string
	companyName?: string
	inn?: number | string
	kpp?: string
	index?: string
	legalAddress?: string
}

/** Строка «Поставщик/Покупатель» по §5 ТЗ. */
export function formatCompanyPartyLine(party: PartyLineFields | undefined): string {
	if (!party?.companyName) return ''
	const type = party.companyType?.trim()
	const shortName = stripLeadingOrgType(party.companyName, type)
	const fullName = type ? `${type} ${shortName}`.trim() : shortName
	const address = formatLegalAddressForParty(party.index, party.legalAddress)
	return [fullName, party.inn, party.kpp, address]
		.filter((part) => part !== '' && part !== undefined && part !== null)
		.join(', ')
}

/** «Получатель» в верхней таблице: тип + название без дубля типа. */
export function formatCompanyRecipientLine(party: Pick<PartyLineFields, 'companyType' | 'companyName'> | undefined): string {
	if (!party?.companyName) return ''
	const type = party.companyType?.trim()
	const shortName = stripLeadingOrgType(party.companyName, type)
	return type ? `${type} ${shortName}`.trim() : shortName
}
